### PR TITLE
upgrade FaceDetectorYN to v2

### PR DESCRIPTION
**Pull request by Wwupup**
Merge with https://github.com/opencv/opencv_extra/pull/1038

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/objdetect/src/face_detect.cpp [hunk 1]
```diff
@@ -115,124 +124,81 @@ class FaceDetectorYNImpl : public FaceDetectorYN
...
+                // Clamp
+                cls_score = MIN(cls_score, 1.f);
+                cls_score = MAX(cls_score, 0.f);
+                obj_score = MIN(obj_score, 1.f);
+                obj_score = MAX(obj_score, 0.f);
+                float score = std::sqrt(cls_score * obj_score);
+                face.at<float>(0, 14) = score;
+
+                // Get bounding box
+                std::vector<Point2f>& priors_layer = priors[i];
```
Comment (fengyuentau) [inline modules/objdetect/src/face_detect.cpp:181]: `std::vector<Point2f>& priors_layer = priors[i];` -> `const std::vector<Point2f>& priors_layer = priors[i];`

File: modules/objdetect/src/face_detect.cpp [hunk 2]
```diff
@@ -37,17 +37,24 @@ class FaceDetectorYNImpl : public FaceDetectorYN
         inputW = input_size.width;
         inputH = input_size.height;
 
+        divisor = 32;
```
Comment (fengyuentau) [inline modules/objdetect/src/face_detect.cpp:40]: I would suggest make it static

File: modules/objdetect/src/face_detect.cpp [hunk 3]
```diff
@@ -37,17 +37,24 @@ class FaceDetectorYNImpl : public FaceDetectorYN
...
+        padW = (int((inputW - 1) / divisor) + 1) * divisor;
+        padH = (int((inputH - 1) / divisor) + 1) * divisor;
+
         scoreThreshold = score_threshold;
         nmsThreshold = nms_threshold;
         topK = top_k;
 
+        strides = {8, 16, 32};
```
Comment (fengyuentau) [inline modules/objdetect/src/face_detect.cpp:48]: I would suggest make it static

**Review by fengyuentau — COMMENTED** [review]
Thanks for the upgrade! I think we need a performance test for this API as well. Could you report the performance difference once it is added?

File: modules/objdetect/src/face_detect.cpp (hunk 1 above)
Comment (Wwupup) [reply modules/objdetect/src/face_detect.cpp:181]: ok

File: modules/objdetect/src/face_detect.cpp (hunk 2 above)
Comment (Wwupup) [reply modules/objdetect/src/face_detect.cpp:40]: Hello, I don't think changing this variable to static is better for performance.

File: modules/objdetect/src/face_detect.cpp (hunk 3 above)
Comment (Wwupup) [reply modules/objdetect/src/face_detect.cpp:48]: Hello, I don't think changing this variable to static is better for performance.

File: modules/objdetect/include/opencv2/objdetect/face.hpp [hunk 4]
```diff
@@ -75,7 +75,7 @@ class CV_EXPORTS_W FaceDetectorYN
     CV_WRAP static Ptr<FaceDetectorYN> create(const String& model,
                                               const String& config,
                                               const Size& input_size,
-                                              float score_threshold = 0.9f,
+                                              float score_threshold = 0.5f,
```
Comment (alalek) [inline modules/objdetect/include/opencv2/objdetect/face.hpp:78]: It is a part of public API.
Changing default values between releases should be avoided as users already expect reliable (0.9) results here.

Comment (Wwupup) [reply modules/objdetect/include/opencv2/objdetect/face.hpp]: Hello, I have restored the original defaults, thanks.

**Comment by asmorkalov** [comment]
@fengyuentau friendly reminder.

**Review by fengyuentau — APPROVED** [review]
LGTM👍

**Comment by fengyuentau** [comment]
> @fengyuentau could you add a note about output format and landmarks too?

Sure, let me do it now.

**Pull request merged**